src/sys/signal.rs

```diff
@@ -471,10 +480,20 @@ pub struct SigAction {
 }
 
 impl SigAction {
-    pub fn new(handler: extern fn(libc::c_int), flags: SockFlag, mask: SigSet) -> SigAction {
+    /// This function will set or unset the flag `SA_SIGINFO` depending on the
+    /// type of the `handler` argument.
+    pub fn new(handler: SigHandler, flags: SockFlag, mask: SigSet) -> SigAction {
         let mut s = unsafe { mem::uninitialized::<sigaction_t>() };
-        s.sa_handler = handler;
-        s.sa_flags = flags;
+        s.sa_handler = match handler {
+            SigHandler::SigDfl => unsafe { mem::transmute(libc::SIG_DFL) },
+            SigHandler::SigIgn => unsafe { mem::transmute(libc::SIG_IGN) },
+            SigHandler::Handler(f) => f,
+            SigHandler::SigAction(f) => unsafe { mem::transmute(f) },
+        };
+        s.sa_flags = match handler {
+            SigHandler::SigAction(_) => flags | SA_SIGINFO,
+            _ => flags - SA_SIGINFO,
```

Comment:
I prefer the minus sign, which represents set difference. And that is exactly what we want to express and achieve here. In addition it is more concise.
I do not think that preventing confusion is a valid argument since the type is stated in the parameter list and (now) by convention we use bitflags for such types.